### DISCUSSION TITLE
Get-PnPRetentionLabel Get-PnPTenantRetentionLabel API Permissions

### DISCUSSION THREAD

**Discussion by anthonywhite** [pnp/powershell#4349] [Q&A]
Hi, does anyone know what API permissions need to be granted to our PnP App Reg to allow `Get-PnPTenantRetentionLabel` command to work? I tried Graph Records Management RW All but still get prompted for Admin consent to something. @reshmee011 ?

**Comment by reshmee011** [comment]
@anthonywhite : Apologies for the late reply. This the permissions I have been prompted to admin consent  
<img width="355" alt="image" src="https://github.com/user-attachments/assets/515cfcf2-f01b-4df9-8367-4a8e23b7e88c">

From Azure Portal the permission "RecordsManagement.Read.All" was granted. 

<img width="813" alt="image" src="https://github.com/user-attachments/assets/63aef196-a4fc-43da-ac9b-55ac6e34b71c">

What permissions are you being prompted? Are they same as the screenshot I provided.

**Reply by anthonywhite** [reply]
hi @reshmee011  - is this a Graph permission? The admin already tried this:  

<img width="346" alt="image" src="https://github.com/user-attachments/assets/2cb68ea3-d37e-4e9a-9cfc-e78a8b98f79b">
 
but still we get a prompt to consent...

**Reply by reshmee011** [reply]
That's all the permissions required, not sure why it's asking about the other permissions like profile. @KoenZomers , @gautamdsheth any ideas why running Get-PnPTenantRetentionLabel prompts for "View your basic profile" and "Maintain access to data you have given it access to"?

**Reply by KoenZomers** [reply]
The `Get-PnPTenantRetentionLabel` cmdlet calls into this Graph API behind the scenes:
https://learn.microsoft.com/en-us/graph/api/security-labelsroot-list-retentionlabel?view=graph-rest-beta&tabs=http

It only requires the delegated permission `RecordsManagement.Read.All` on Microsoft Graph and will not work under an application context.

The entire permission requesting and validation in PnP PowerShell I'm currently rewriting to make it much more clear which permissions are needed and which APIs are being called into. In the current latest version some stuff of that is already baked in when using `-Verbose` with your cmdlet execution. In the next version, there will be much more of it. Sneak preview:

![image](https://github.com/user-attachments/assets/ccad2764-acc7-4850-90ca-9e7fdef8ae41)

**Reply by anthonywhite** [reply]
@KoenZomers @reshmee011 thanks for the replies. I will ask our admin to add that specific scope as it looks like the `ReadWrite.All` that we have is not matching. thx

**Reply by anthonywhite** [reply]
...and having added that scope, running with -Verbose is now showing this cryptic error message

```
VERBOSE: Making GET call to https://graph.microsoft.com/beta/security/labels/retentionLabels
VERBOSE: Response failed with HTTP 500 InternalServerError containing 448 characters: {"error":{"code":"UnknownError","message":"{\"ErrorCode\":\"UnknownError\",\"Message\":\"Failed to contact DataInsights in EOP - Tenant = 8ab286da-9dbb-431b-b4eb-360898848943. Exception = DataInsights command(GET) FAILED - Forbidden. TargetServer = CWXP265MB3831.GBRP265.PROD.OUTLOOK.COM\"}","innerError":{"date":"2024-10-03T08:36:42","request-id":"b2d6c671-5abd-4b49-a66a-cbfab4f2aebe","client-request-id":"b2d6c671-5abd-4b49-a66a-cbfab4f2aebe"}}}
Get-PnPTenantRetentionLabel: Internal Server Error (500): {"ErrorCode":"UnknownError","Message":"Failed to contact DataInsights in EOP - Tenant = 8ab286da-9dbb-431b-b4eb-360898848943. Exception = DataInsights command(GET) FAILED - Forbidden. TargetServer = CWXP265MB3831.GBRP265.PROD.OUTLOOK.COM"}
```

any ideas what might cause this?
